File: api/healthcheck/HealthCheck.go
package healthcheck

import (
	"gopkg.in/validator.v2"
)

// Node node in the
type HealthCheck struct {
	ID       string `json:"id" validate:"nonzero"`
	Name     string `json:"name" validate:"nonzero"`
	Resource string `json:"resource" validate:"nonzero"`
	Status   string `json:"status" validate:"nonzero"`
	Message  string `json:"message" validate:"nonzero"`
}

type NodeHealthCheck struct {
	Hostname string `json:"hostname" validate:"nonzero"`
	ID       string `json:"id" validate:"nonzero"`
	Status   string `json:"status" validate:"nonzero"`
}

type Node struct {
	Hostname     string        `json:"hostname" validate:"nonzero"`
	ID           string        `json:"id" validate:"nonzero"`
	HealthChecks []HealthCheck `json:"healthchecks" validate:"nonzero"`
}

func (s HealthCheck) Validate() error {

	return validator.Validate(s)
}

// health/nodes

File: pyclient/zeroos/orchestrator/sal/healthcheck.py
import os
from zeroos.core0.client.client import Timeout
import json
import hashlib


class ContainerContext:
    def __init__(self, node, flist):
        self.node = node
        self.flist = flist
        self.container = None
        self._name = 'healthcheck_{}'.format(hashlib.md5(flist.encode()).hexdigest())

    def __enter__(self):
        try:
            self.container = self.node.containers.get(self._name)
        except LookupError:
            self.container = self.node.containers.create(self._name, self.flist, host_network=True, privileged=True)
        return self.container

    def __exit__(self, exc_type, exc_val, exc_tb):
        return


class HealthCheck:
    def __init__(self, node):
        self.node = node
        self.healtcheckfolder = os.path.join(os.path.abspath(os.path.dirname(__file__)), 'healthchecks')

    def with_container(self, flist):
        return ContainerContext(self.node, flist)

    def run(self, container, name, timeout=None):
        try:
            healthcheckfile = os.path.join(self.healtcheckfolder, name + '.py')
            if not os.path.exists(healthcheckfile):
                raise RuntimeError("Healtcheck with name {} not found".format(name))
            container.client.filesystem.upload_file('/tmp/{}.py'.format(name), healthcheckfile)
            try:
                job = container.client.bash('python3 /tmp/{}.py'.format(name))
                response = job.get(timeout)
            except Timeout:
                container.client.job.kill(job.id, 9)
                raise RuntimeError("Failed to execute {} on time".format(name))
            if response.state == 'ERROR':
                raise RuntimeError("Failed to execute {} {}".format(name, response.stdout))
            try:
                return json.loads(response.stdout)
            except Exception:
                raise RuntimeError("Failed to parse response of {}".format(name))
        except Exception as e:
            healtcheck = {
                'id': name,
                'resource': '',
                'status': 'ERROR',
                'message': str(e)
            }
            return [healtcheck]

    def calc_cpu_mem(self):
        from .healthchecks.cpu_mem_core_check import action
        results = action(self.node)
        return results
